README.md
# 📷+🗺+⚡️ = CamMap

CamMap is a swift library that helps apps take photos and grab locations at the same time.

![Sample screenshot](Images/ss1.jpeg?raw=true)

## Features 🚀

- [x] Capture images as JPEG
- [x] Use the current location as soon as possible.
- [x] Floating pin for better selecting the coordinates. Tap to save the location.
- [x] Drag and drop pin within the map.
- [x] Store photos in the device. Provide a configurable flag to turn it off.
- [x] Handle portrait and landscape nicely.

# Installation 💻

### Using [CocoaPods](http://cocoapods.org/)

Add `pod 'CamMap'` to your `Podfile` and run `pod install`. Also add `use_frameworks!` to the `Podfile`.

```
use_frameworks!
pod 'CamMap'
```

### Usage

Import it first:

```
import CamMap
```

Then implement the delegate protocol:

```
extension YouViewController: CamMapDelegate {
    func camMapDidComplete(images: [UIImage], location: CLLocationCoordinate2D?) {
        // This method gets called if the user took at least one pic
    }

    func camMapDidCancel() {
        // Or this gets called if the user cancelled the action
    }

    func camMapHadFailure(error: CamMapError) {
        // Is called when there's a non recoverall issue with the camera or location maanger
    }

    func camMapPermissionFailure(type: PermType, details: String) {
        // Gets call if there is a missing permission
    }
}
```

And within a method from YouViewController present the view controller:

```
let vc = CamMapViewController()
vc.delegate = self
present(vc, animated: true, completion: nil)
```

## Permissions

You need to add the following elements to your Info.plist file, oherwise the app will crash with a related message.

- NSLocationWhenInUseUsageDescription
- NSCameraUsageDescription
- NSPhotoLibraryAddUsageDescription

## Code sample

You can find a simple example within this repo. To try it please clone it and then execute:

```
cd Examples/SimpleCamMapExample/
pod install
open SimpleCamMapExample.xcworkspace
```

# Roadmap 🏁

- [x] Better action buttons
- [x] Indicate how many photos were taken
- [ ] Be able to cancel/close the controller after one photo got in
- [ ] Customizable map and camera sizes
- [ ] Be able to delete selected photos
- [ ] Better permission handling

# Author

Humber Aquino <humber@ogahunt.com> [@goku2](https://twitter.com/goku2)

# License

CamMap is released under the MIT license.
See LICENSE for details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.2] - 2018-11-09

### Changed

-   Fix crash when opening the controller from the simulator

## [0.0.1] - 2018-10-27

### Added

-   Initial version


CamMap/ImageCountView.swift
//
//  ImageCountView.swift
//  CamMap
//
//  Created by Humberto Aquino on 10/27/18.
//  Copyright © 2018 Humberto Aquino. All rights reserved.
//

import UIKit

class ImageCountView: UIView {
    var originalImage: UIImage?
    var previewImageView: UIImageView!
    var currentCount: Int = 0
    var countLabel: UILabel!

    convenience init() {
        self.init(frame: CGRect.zero)
        render()
    }

    private func render() {
        previewImageView = UIImageView()
        previewImageView.layer.cornerRadius = 5.0
        previewImageView.clipsToBounds = true

        // Core attrs
        countLabel = UILabel()
        countLabel.textColor = UIColor.white
        countLabel.font = UIFont.boldSystemFont(ofSize: 23)
        countLabel.textAlignment = .center

        // Shadow
        countLabel.layer.shadowColor = UIColor.black.cgColor
        countLabel.layer.shadowRadius = 3.0
        countLabel.layer.shadowOpacity = 0.8
        countLabel.layer.shadowOffset = .zero
        countLabel.layer.masksToBounds = false

        addSubview(previewImageView)
        addSubview(countLabel)

        bringSubviewToFront(countLabel)
    }

    // Core method to interact with the view's state

    func updateWith(image: UIImage, count: Int) {
        originalImage = image
        currentCount = count

        previewImageView.frame = bounds

        let width: CGFloat = 40
        let height: CGFloat = 30
        let x = bounds.width / 2 - width / 2
        let y = bounds.height / 2 - height / 2
        countLabel.frame = CGRect(x: x, y: y, width: width, height: height)

        previewImageView.image = image
        countLabel.text = "\(count)"
    }
}


CamMapTests/CamMapTests.swift
//
//  CamMapTests.swift
//  CamMapTests
//
//  Created by Humberto Aquino on 10/26/18.
//  Copyright © 2018 Humberto Aquino. All rights reserved.
//

import XCTest
@testable import CamMap

class CamMapTests: XCTestCase {

    override func setUp() {
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }

    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
    }

    func testExample() {
        // This is an example of a functional test case.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }

    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure {
            // Put the code you want to measure the time of here.
        }
    }

}


Examples/SimpleCamMapExample/SimpleCamMapExample/AppDelegate.swift
//
//  AppDelegate.swift
//  SimpleCamMapExample
//
//  Created by Humberto Aquino on 10/26/18.
//  Copyright © 2018 Humberto Aquino. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



Examples/SimpleCamMapExample/SimpleCamMapExample/ViewController.swift
//
//  ViewController.swift
//  SimpleCamMapExample
//
//  Created by Humberto Aquino on 10/26/18.
//  Copyright © 2018 Humberto Aquino. All rights reserved.
//

import UIKit
import SnapKit
import CamMap
import CoreLocation

class ViewController: UIViewController {

    var openButton: UIButton!

    override func viewDidLoad() {
        super.viewDidLoad()

        openButton = UIButton(type: .system)
        openButton.setTitle("Open", for: .normal)
        openButton.tintColor = UIColor.white
        openButton.addTarget(self, action: #selector(openCamMap(_:)), for: .touchUpInside)

        view.addSubview(openButton)

        openButton.snp.makeConstraints { make in
            make.center.equalTo(self.view)
            make.width.equalTo(100)
            make.height.equalTo(40)
        }
        view.backgroundColor = UIColor.darkGray
    }

    @objc
    func openCamMap(_: Any) {
        let vc = CamMapViewController()
        vc.delegate = self
        present(vc, animated: false, completion: nil)
    }

}

extension ViewController: CamMapDelegate {
    func camMapDidComplete(images: [UIImage], location: CLLocationCoordinate2D?) {
        print("Completed. Got \(images.count) images and Location: \(String(describing: location))")
        self.dismiss(animated: false, completion: nil)
    }

    func camMapDidCancel() {
        print("Cancelled")
        self.dismiss(animated: false, completion: nil)
    }

    func camMapPermissionFailure(type: PermType, details: String) {
        print("Permission failure: \(details)")
        self.dismiss(animated: false, completion: nil)
    }

    func camMapHadFailure(error: CamMapError) {
        print("Unexpected error: \(error)")
        self.dismiss(animated: false, completion: nil)
    }

}


CamMap/CamMapViewController.swift
//
//  CamMapViewController.swift
//  CamMap
//
//  Created by Humberto Aquino on 10/26/18.
//  Copyright © 2018 Humberto Aquino. All rights reserved.
//

import AVFoundation
import Foundation
import MapKit
import UIKit

public protocol CamMapDelegate: class {
    func camMapDidComplete(images: [UIImage], location: CLLocationCoordinate2D?)
    func camMapDidCancel()
    func camMapHadFailure(error: CamMapError)
    func camMapPermissionFailure(type: PermType, details: String)
}

public enum CamMapError: Error {
    case noVideo
    case cantStartCapture(Error)
    case locationFailed(Error)
    case photoJPEGFailed
    case imageDataTransformation
    case cantSaveImage(Error?)
}

public enum PermType {
    case location
    case camera
    case photos
}

public class CamMapViewController: UIViewController {
    // Params
    public var regionRadius: CLLocationDistance = 400
    public var shouldStorePhotos = true

    // Map and Location
    var locationManager = CLLocationManager()
    var mapView: MKMapView!
    var zoomLevel: Float = 14.0
    var selectedCoordinate: CLLocationCoordinate2D?
    var centerMapPin: UIImageView!

    // Camera
    var previewView: UIView!
    var captureSession: AVCaptureSession!
    var videoPreviewLayer: AVCaptureVideoPreviewLayer!
    var capturePhotoOutput: AVCapturePhotoOutput!
    var images: [UIImage] = []

    // Actions
    var captureButton: UIButton!
    var completeButton: UIButton!
    var imageCountView: ImageCountView!

    // Layout
    var protraitConstrains: [NSLayoutConstraint]!
    var landscapeConstrains: [NSLayoutConstraint]!

    // Delegate
    public weak var delegate: CamMapDelegate?

    // MARK: - Life cycle

    public override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
        setupLocationManager()
    }

    public override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        updateVideoPreviewLayerSize()

        // Roundness
        captureButton.layer.cornerRadius = 0.5 * captureButton.bounds.size.width
        captureButton.clipsToBounds = true
        captureButton.backgroundColor = UIColor.gray
        captureButton.alpha = 0.9

        let cameraImage = ImageUtils.loadImage(named: "camera")
        captureButton.setImage(cameraImage, for: .normal)

        // Shadow
        captureButton.layer.shadowColor = UIColor.black.cgColor
        captureButton.layer.shadowOffset = CGSize(width: 0.0, height: 0.0)
        captureButton.layer.masksToBounds = false
        captureButton.layer.shadowRadius = 5.0
        captureButton.layer.shadowOpacity = 0.3
    }

    public override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        // Only request is location is not selected
        if selectedCoordinate != nil {
            return
        }

        locationManager.requestWhenInUseAuthorization()
        locationManager.requestLocation()
        if let location = locationManager.location {
            centerMapOnLocation(location: location)
        }
    }

    public override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransition(to: size, with: coordinator)
        // Update camera orientation
        let videoOrientation: AVCaptureVideoOrientation
        switch UIDevice.current.orientation {
        case .portrait:
            videoOrientation = .portrait
        case .portraitUpsideDown:
            videoOrientation = .portraitUpsideDown
        case .landscapeLeft:
            videoOrientation = .landscapeRight
        case .landscapeRight:
            videoOrientation = .landscapeLeft
        default:
            videoOrientation = .portrait
        }
        videoPreviewLayer.connection?.videoOrientation = videoOrientation

        // Update view constraints
        updateLayoutConstraints()
    }

    // MARK: - Setup UI

    func setupUI() {
        setupMap()
        setupCamera()
        setupButtons()

        configureViewsAndConstrains()
        updateLayoutConstraints()
    }

    func setupMap() {
        // Map
        mapView = MKMapView()
        mapView.delegate = self
        mapView.showsUserLocation = true

        let image = ImageUtils.loadImage(named: "drop-pin")
        centerMapPin = UIImageView(image: image)
        centerMapPin.isUserInteractionEnabled = true
        let tapRecognizer = UITapGestureRecognizer(target: self, action: #selector(centerMapPinTapped(_:)))
        centerMapPin.addGestureRecognizer(tapRecognizer)
    }

    func setupCamera() {
        // Camera
        previewView = UIView()
        guard let captureDevice = AVCaptureDevice.default(for: AVMediaType.video) else {
            delegate?.camMapHadFailure(error: CamMapError.noVideo)
            return
        }

        do {
            let input = try AVCaptureDeviceInput(device: captureDevice)

            captureSession = AVCaptureSession()
            captureSession.addInput(input)

            videoPreviewLayer = AVCaptureVideoPreviewLayer(session: captureSession)
            videoPreviewLayer.videoGravity = .resizeAspectFill
            previewView.layer.addSublayer(videoPreviewLayer)

            captureSession.startRunning()

            // Get an instance of ACCapturePhotoOutput
            capturePhotoOutput = AVCapturePhotoOutput()
            capturePhotoOutput.isHighResolutionCaptureEnabled = true
            // Set the output on the capture session
            captureSession.addOutput(capturePhotoOutput)
        } catch {
            delegate?.camMapHadFailure(error: CamMapError.cantStartCapture(error))
        }
    }

    func setupButtons() {
        // Buttons
        captureButton = UIButton(type: .custom)
        captureButton.addTarget(self, action: #selector(takePicture(_:)), for: .touchUpInside)

        completeButton = UIButton(type: .system)
        completeButton.setTitle("Cancel", for: .normal)
        completeButton.addTarget(self, action: #selector(completeAction(_:)), for: .touchUpInside)

        imageCountView = ImageCountView()
    }

    func configureViewsAndConstrains() {
        view.addSubview(centerMapPin)
        view.addSubview(mapView)
        view.addSubview(previewView)
        view.addSubview(captureButton)
        view.addSubview(completeButton)
        view.addSubview(imageCountView)

        view.bringSubviewToFront(centerMapPin)

        mapView.translatesAutoresizingMaskIntoConstraints = false
        previewView.translatesAutoresizingMaskIntoConstraints = false
        captureButton.translatesAutoresizingMaskIntoConstraints = false

        centerMapPin.translatesAutoresizingMaskIntoConstraints = false
        completeButton.translatesAutoresizingMaskIntoConstraints = false
        imageCountView.translatesAutoresizingMaskIntoConstraints = false

        protraitConstrains = [
            mapView.topAnchor.constraint(equalTo: self.view.topAnchor),
            mapView.leadingAnchor.constraint(equalTo: self.view.leadingAnchor),
            mapView.trailingAnchor.constraint(equalTo: self.view.trailingAnchor),
            mapView.heightAnchor.constraint(equalTo: self.view.heightAnchor, multiplier: 0.30),

            previewView.topAnchor.constraint(equalTo: self.mapView.bottomAnchor),
            previewView.leftAnchor.constraint(equalTo: self.view.leftAnchor),
            previewView.rightAnchor.constraint(equalTo: self.view.rightAnchor),
            previewView.bottomAnchor.constraint(equalTo: self.view.bottomAnchor),

            centerMapPin.centerXAnchor.constraint(equalTo: self.mapView.centerXAnchor),
            centerMapPin.centerYAnchor.constraint(equalTo: self.mapView.centerYAnchor),
            centerMapPin.widthAnchor.constraint(equalToConstant: 35),
            centerMapPin.heightAnchor.constraint(equalToConstant: 35),

            captureButton.bottomAnchor.constraint(equalTo: self.view.bottomAnchor, constant: -30),
            captureButton.widthAnchor.constraint(equalToConstant: 60),
            captureButton.heightAnchor.constraint(equalToConstant: 60),
            captureButton.centerXAnchor.constraint(equalTo: self.view.centerXAnchor),

            completeButton.centerYAnchor.constraint(equalTo: self.captureButton.centerYAnchor),
            completeButton.heightAnchor.constraint(equalToConstant: 35),
            completeButton.rightAnchor.constraint(equalTo: self.view.rightAnchor, constant: -40),
            completeButton.leftAnchor.constraint(equalTo: self.captureButton.rightAnchor, constant: 40),

            imageCountView.centerYAnchor.constraint(equalTo: self.captureButton.centerYAnchor),
            imageCountView.widthAnchor.constraint(equalToConstant: 50),
            imageCountView.heightAnchor.constraint(equalToConstant: 50),
            imageCountView.leftAnchor.constraint(equalTo: self.view.leftAnchor, constant: 55),
        ]

        landscapeConstrains = [
            mapView.topAnchor.constraint(equalTo: self.view.topAnchor),
            mapView.leadingAnchor.constraint(equalTo: self.view.leadingAnchor),
            mapView.bottomAnchor.constraint(equalTo: self.view.bottomAnchor),
            mapView.widthAnchor.constraint(equalTo: self.view.widthAnchor, multiplier: 0.30),

            previewView.topAnchor.constraint(equalTo: self.view.topAnchor),
            previewView.leftAnchor.constraint(equalTo: self.mapView.rightAnchor),
            previewView.rightAnchor.constraint(equalTo: self.view.rightAnchor),
            previewView.bottomAnchor.constraint(equalTo: self.view.bottomAnchor),

            centerMapPin.centerXAnchor.constraint(equalTo: self.mapView.centerXAnchor),
            centerMapPin.centerYAnchor.constraint(equalTo: self.mapView.centerYAnchor),
            centerMapPin.widthAnchor.constraint(equalToConstant: 35),
            centerMapPin.heightAnchor.constraint(equalToConstant: 35),

            captureButton.rightAnchor.constraint(equalTo: self.view.rightAnchor, constant: -30),
            captureButton.widthAnchor.constraint(equalToConstant: 60),
            captureButton.heightAnchor.constraint(equalToConstant: 60),
            captureButton.centerYAnchor.constraint(equalTo: self.view.centerYAnchor),

            completeButton.centerXAnchor.constraint(equalTo: self.captureButton.centerXAnchor),
            completeButton.widthAnchor.constraint(equalToConstant: 100),
            completeButton.heightAnchor.constraint(equalToConstant: 35),
            completeButton.topAnchor.constraint(equalTo: self.captureButton.bottomAnchor, constant: 40),

            imageCountView.centerXAnchor.constraint(equalTo: self.captureButton.centerXAnchor),
            imageCountView.widthAnchor.constraint(equalToConstant: 50),
            imageCountView.heightAnchor.constraint(equalToConstant: 50),
            imageCountView.bottomAnchor.constraint(equalTo: self.captureButton.topAnchor, constant: -40),
        ]
    }

    func updateLayoutConstraints() {
        let portrait = UIDevice.current.orientation != .landscapeLeft && UIDevice.current.orientation != .landscapeRight
        if portrait {
            NSLayoutConstraint.deactivate(landscapeConstrains)
            NSLayoutConstraint.activate(protraitConstrains)
        } else {
            NSLayoutConstraint.deactivate(protraitConstrains)
            NSLayoutConstraint.activate(landscapeConstrains)
        }
    }

    // MARK: - Actions

    @objc
    func takePicture(_: Any) {
        // First animate the action
        UIView.animate(withDuration: 0.2, animations: {
            self.captureButton.transform = CGAffineTransform(scaleX: 0.8, y: 0.8)
        }, completion: { _ in
            UIView.animate(withDuration: 0.2) {
                self.captureButton.transform = CGAffineTransform.identity
            }
        })

        // Make sure capturePhotoOutput is valid
        guard let capturePhotoOutput = self.capturePhotoOutput else { return }
        // Get an instance of AVCapturePhotoSettings class
        let photoSettings = AVCapturePhotoSettings()
        // Set photo settings for our need
        photoSettings.isAutoStillImageStabilizationEnabled = true
        photoSettings.isHighResolutionPhotoEnabled = true
        photoSettings.flashMode = .off
        // Call capturePhoto method by passing our photo settings and a
        // delegate implementing AVCapturePhotoCaptureDelegate
        capturePhotoOutput.capturePhoto(with: photoSettings, delegate: self)
    }

    @objc
    func completeAction(_: Any) {
        if images.count == 0 {
            delegate?.camMapDidCancel()
        } else {
            let selectedLocation = identifySelectedLocation()
            delegate?.camMapDidComplete(images: images, location: selectedLocation)
        }
    }

    @objc
    func centerMapPinTapped(_: UIGestureRecognizer) {
        locationManager.stopUpdatingLocation()

        // Get the center of the map in coordiantes
        let center = mapView.centerCoordinate

        // Remove the image
        centerMapPin.isUserInteractionEnabled = false
        centerMapPin.isHidden = true

        // Place a Marker
        selectedCoordinate = center

        let annotation = MKPointAnnotation()

        annotation.coordinate = selectedCoordinate!
        mapView.addAnnotation(annotation)

        centerMapOnLocation(location: CLLocation(latitude: center.latitude, longitude: center.longitude))
    }

    // MARK: - Utils

    func identifySelectedLocation() -> CLLocationCoordinate2D {
        guard let selectedCoordinate = self.selectedCoordinate else {
            return mapView.centerCoordinate
        }
        return selectedCoordinate
    }

    func updateVideoPreviewLayerSize() {
        if let videoPreviewLayer = videoPreviewLayer {
            videoPreviewLayer.frame = previewView.bounds
        }
    }

    func markCompleteState() {
        completeButton.setTitle("Done", for: .normal)
    }
}

// MARK: - AVCapturePhotoCaptureDelegate

extension CamMapViewController: AVCapturePhotoCaptureDelegate {
    public func photoOutput(_: AVCapturePhotoOutput, didFinishProcessingPhoto photo: AVCapturePhoto, error _: Error?) {
        guard let imageData = photo.fileDataRepresentation() else {
            delegate?.camMapHadFailure(error: CamMapError.photoJPEGFailed)
            return
        }

        let capturedImage = UIImage(data: imageData, scale: 1.0)
        guard let image = capturedImage else {
            delegate?.camMapHadFailure(error: CamMapError.imageDataTransformation)
            return
        }

        images.append(image)
        markCompleteState()

        imageCountView.updateWith(image: image, count: images.count)

        if shouldStorePhotos {
            UIImageWriteToSavedPhotosAlbum(image, self, #selector(image(_:didFinishSavingWithError:contextInfo:)), nil)
        }
    }

    @objc func image(_: UIImage, didFinishSavingWithError error: NSError?, contextInfo _: UnsafeRawPointer) {
        if let error = error {
            delegate?.camMapHadFailure(error: CamMapError.cantSaveImage(error as Error))
            return
        }

        // TODO: Show some indication that the image got saved
    }
}

// MARK: - CLLocationManagerDelegate

extension CamMapViewController: CLLocationManagerDelegate {
    func setupLocationManager() {
        locationManager = CLLocationManager()
        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
        locationManager.distanceFilter = 50
    }

    public func locationManager(_: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        let location: CLLocation = locations.last!
        centerMapOnLocation(location: location)
        locationManager.stopUpdatingLocation()
    }

    // Handle authorization for the location manager
    public func locationManager(_: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        switch status {
        case .restricted:
            delegate?.camMapPermissionFailure(type: .location, details: "Location access was restricted.")
        case .denied:
            delegate?.camMapPermissionFailure(type: .location, details: "User denied access to location.")
            mapView.isHidden = false
        case .notDetermined:
            delegate?.camMapPermissionFailure(type: .location, details: "Location status not determined.")
        default:
            // No op
            break
        }
    }

    // Handle location manager errors.
    public func locationManager(_: CLLocationManager, didFailWithError error: Error) {
        locationManager.stopUpdatingLocation()
        delegate?.camMapHadFailure(error: CamMapError.locationFailed(error))
    }

    func centerMapOnLocation(location: CLLocation) {
        let coordinateRegion = MKCoordinateRegion(center: location.coordinate,
                                                  latitudinalMeters: regionRadius, longitudinalMeters: regionRadius)
        mapView.setRegion(coordinateRegion, animated: true)
    }

    func replaceMarkerWithSelectedCoordiante() {
        guard let selectedCoordinate = selectedCoordinate else { return }

        mapView.removeAnnotations(mapView.annotations)

        let annotation = MKPointAnnotation()
        annotation.coordinate = selectedCoordinate
        mapView.addAnnotation(annotation)
    }
}

// MARK: - MKMapViewDelegate

extension CamMapViewController: MKMapViewDelegate {
    public func mapView(_: MKMapView, annotationView view: MKAnnotationView, didChange newState: MKAnnotationView.DragState, fromOldState _: MKAnnotationView.DragState) {
        switch newState {
        case .starting:
            view.dragState = .dragging
        case .ending, .canceling:
            view.dragState = .none
        default: break
        }
    }

    public func mapView(_: MKMapView, viewFor annotation: MKAnnotation) -> MKAnnotationView? {
        if annotation is MKPointAnnotation {
            let pinAnnotationView = MKPinAnnotationView(annotation: annotation, reuseIdentifier: "locationPin")

            pinAnnotationView.isDraggable = true
            pinAnnotationView.canShowCallout = true
            pinAnnotationView.animatesDrop = true

            return pinAnnotationView
        }

        return nil
    }
}
